R/model_builder.R

```diff
@@ -23,78 +34,144 @@ do_data <- function(funcname) {
+build_var.test <- build_data("var.test")
 
 #' kmeans wrapper with do
 #' @param centers Set an integer number to decide how many clusters (groups) to build.
 #' @param keep.source It will make .source.data column to preserve source data.
 #' @param seed This is random seed. You can change the result if you change this number.
 #' @return deta frame which has kmeans model
 #' @export
-do_kmeans <- function(df, ..., centers=3, keep.source = FALSE, seed=0){
+build_kmeans.kv <- function(df,
+                            subject,
+                            key,
+                            value,
+                            centers=3,
+                            iter.max = 10,
+                            nstart = 1,
+                            algorithm = "Hartigan-Wong",
+                            trace = FALSE,
+                            keep.source = FALSE,
+                            seed=0,
+                            augment=F,
+                            fun.aggregate=mean,
+                            fill=0){
   loadNamespace("dplyr")
+  loadNamespace("lazyeval")
+  loadNamespace("tidyr")
+  loadNamespace("broom")
   set.seed(seed)
-  labels <- attr(df, "labels")
-  if(is.null(labels)){
-    grouped_cname = NULL
-  } else {
-    grouped_cname = colnames(labels)
+
+  row_col <- col_name(substitute(subject))
+  col_col <- col_name(substitute(key))
+  value_col <- col_name(substitute(value))
+
+  build_kmeans_each <- function(df){
+    mat <- simple_cast(df, row_col, col_col, value_col, fun.aggregate = fun.aggregate, fill=fill)
+    kmeans_ret <- kmeans(mat, centers = centers, iter.max = 10, nstart = nstart, algorithm = algorithm, trace = trace)
+    if(augment){
+      if(length(kmeans_ret$cluster) == nrow(df)){
+        broom::augment(kmeans_ret, df)
+      } else {
+        broom::augment(kmeans_ret, setNames(data.frame(rownames(mat),stringsAsFactors = F), row_col))
+      }
+    } else {
+      kmeans_ret
+    }
   }
-  tryCatch({
-    selected_df <- dplyr::select(df, ...)
-    # check if dots are all about column selection
-    if(!all(colnames(selected_df) %in% colnames(df))){
-      # columns which were not in original data frame
-      no_column <- colnames(selected_df)[!colnames(selected_df) %in% colnames(df)]
-      stop(paste(no_column, "is undefined in the data frame and argument", collapse = " "))
+  grouped_column <- grouped_by(df)
+  model_column <- avoid_conflict(grouped_column, "model")
+  source_column <- avoid_conflict(grouped_column, "source.data")
+
+  if(keep.source){
+    output <- (
+      df
+      %>%  dplyr::do_(.dots=setNames(list(~build_kmeans_each(.), ~(.)), c(model_column, source_column)))
+    )
+    # Add a class for Exploratyry to recognize the type of .source.data
+    class(output[[source_column]]) <- c("list", ".source.data")
+  } else {
+    output <- (
+      df
+      %>%  dplyr::do_(.dots=setNames(list(~build_kmeans_each(.)), model_column))
+    )
+    if(augment){
```

Comment:
Yes, augment should be out from there.
I'll change the first condition to `keep.source | !augment` and put out the augment condition from the else block